### ISSUE TITLE
PDF library for Invoices

### ISSUE THREAD

**Issue by arnisjuraga** [copona/copona#137]
Hi!

We need to generate and attach invoices as PDF documents.
I have used https://tcpdf.org library for this on a project.

Any objections extending Copona invoices as PDF?

**Comment by prhost** [comment]
Copona already generates incoives in PDF or something in PDF?

**Comment by arnisjuraga** [comment]
No, not yet. But I have integrated tcpf in Opencart 2, so - will migrate it to Copona. Let's it be Tcpdf? Available for composer I believe.

**Comment by prhost** [comment]
I think it would not be a good idea to leave this fixed in the core of the copona, this could be an extension to part. But if I decide to put it I think these other libraries would be better.

see others libraries:
https://github.com/mpdf/mpdf
https://github.com/dompdf/dompdf

**Comment by arnisjuraga** [comment]
I have exeprience with dompdf. "Quite" problematic.

2017. gada 24. maijs 19:55, "Kallef Alexandre de Souza" <
notifications@github.com> rakstīja:

I think it would not be a good idea to leave this fixed in the core of the
copona, this could be an extension to part. But if I decide to put it I
think these other libraries would be better.

see others libraries:
https://github.com/mpdf/mpdf
https://github.com/dompdf/dompdf

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/copona/copona/issues/137#issuecomment-303785828>, or mute
the thread

**Comment by arnisjuraga** [comment]
Also - PDF has become as "must have" and lack of support in core is big drawback IMHO

**Comment by prhost** [comment]
But it is not easier to leave this as an option for the developer, which library he likes to use? It's just him performing a composer requires something that will already be available to him. Do you think dompdf is problematic, others are not, if it is put in the core of the copona, then at least we should make an interface or abstract class and compel the developer to extend it, what do you think?

In addition to the PDF library, do you intend to make a button to download the invoice inside the panel, something like this? Or just make the library available?

**Comment by RaidoDev** [comment]
> But it is not easier to leave this as an option for the developer, which library he likes to use? It's just him performing a composer requires something that will already be available to him. Do you think dompdf is problematic, others are not, if it is put in the core of the copona, then at least we should make an interface or abstract class and compel the developer to extend it, what do you think?

Yes, this is what I have been thinking - wrapping PDF class would be the best solution, similar to cache or SQL engine. Just more complicated to implement, but that would be the correct way. 

> In addition to the PDF library, do you intend to make a button to download the invoice inside the panel, something like this? Or just make the library available?

Yes of course. And pdf is sent as an attachment to the customer upon purchase, instead of bulk HTML mail,

**Comment by prhost** [comment]
With tcpdf know what size of the PDF file generated to attach in the request? Has a library that makes the PDF file very heavy, especially with html, to make a class involving the library, and give the option in the administrator settings to enable or not to send via html or via PDF.

**Comment by prhost** [comment]
Sometimes it is not very difficult, to make a method to pass options, maybe even in the constructor, plus some others like render, output, save, there we do an integration with this tcpdf to serve as an example, and load it into the registry, the developer You want to overwrite the library, you need to use the same interface and give a registry set on the previous pdf library.

**Comment by stefanhuber** [comment]
PDF is quite difficult to get right, as a store owner probably wants to have many customizations to the generated PDF, e.g. integrating his corporate identity.

I know from Magento that it is very hard to change the styling of the PDF. This needs to be available to the developer very easily I think.

I use TCPDF very often the HTML -> PDF conversion works very well. I must say that I don't know any other PDF library.

IMHO the best approach would be to have a html template, which is converted to PDF, than developers can modify it easily as part of the theme.

**Comment by prhost** [comment]
@stefanhuber converter html para pdf com o TCPDF não fica pesado? já tive alguns problemas assim, só não recordo qual library eu usei.

**Comment by arnisjuraga** [comment]
Some reviews, votes:

https://stackoverflow.com/questions/1648715/which-one-is-the-best-pdf-api-for-php
http://plavatvornica.com/how-to-efficiently-transfer-html-to-pdf/

I am already running tcpdf on the OPencart project, seem to be working great!

**Comment by prhost** [comment]
It seems that DOMPDF is good according to the benchmark, only with long html it bursts the memory.

**Comment by prhost** [comment]
But I think this would not be a problem, if you interface the platform and the developer, he can use whatever libary PDF he wants, we can even make adapters for the 3 if he has the time.

**Comment by landsman** [comment]
Hello guys,
I am working on it here: https://github.com/BurdaPraha/oc_invoice_pdf

**Comment by arnisjuraga** [comment]
@landsman your solution will work only with DomPDF?

**Comment by landsman** [comment]
@arnisjuraga yeah I hope. I want solve SVG rendering (my logo is in SVG: https://github.com/BurdaPraha/oc_svg) and sending with emails for now.

**Comment by arnisjuraga** [comment]
I have this done witvh TCPDF, write me an email, i can share the code

**Comment by landsman** [comment]
I send it.

**Comment by landsman** [comment]
@arnisjuraga I was write to info@copona.org, right?

**Comment by arnisjuraga** [comment]
Yes. Received.

**Comment by landsman** [comment]
@arnisjuraga but without reply :)))

**Comment by guntisluta** [comment]
@landsman have you succeed with PDF integration? I could use that and we can work together to integrate it into Copona?

**Comment by arnisjuraga** [comment]
Yes, it seems, we need to do this. @landsman @stefanhuber @guntisluta  and @prhost  - 
1) I would vote for tcpdf library, @guntisluta seems to have good results on that. 
I have worked with dompdf also. Even though they both dosn't seems to support ```float:``` properly.

2) library should be installable by ```composer```.

The best way would be to make some wrapper in Copona, to be called like ```->makePDF``` method on data and template. Which takes existing PDF library enabled (as we have discussed it before). PDF should be something included in Copona core for sure!

If someone has any progress on this, please share!
